### ISSUE TITLE
False alarms for Redis and Memcache

### ISSUE THREAD

**Issue by jurgenhaas** [firehol/netdata#999]
On some of our hosts we're rarely getting false alarms for Redis and Memcached. Just now I came across a host where Redis server is running but no application is using it. But it is available and is just sitting there being idle.

NetData is reporting that no data was collected for more than 61 seconds. I wonder what's going wrong. Is NetData monitoring if there is some activity with Redis and alerts if there is none or does NetData access Redis itself and reporting a problem if itself can't receive anything?

**Comment by ktsaou** [comment]
I have also seen this once.
If you look at the chart, you will notice that although it is collected, it is this number of seconds behind. It lost some time for some reason, but when the values were finally collected it did not catch up.

@paulfantom this kind of timing is controlled by the number of microseconds since the last successful collection, sent at the `BEGIN` lines.

If for any reason, data are not send to netdata, this number should not be reset. The next successful collection should include all the time since last `BEGIN` line of the same chart.

I haven't verified this, but this is the only reason I can think of.

@paulfantom can you check the python code?

**Comment by ktsaou** [comment]
I marked it as bug, since either way (python code bug or not), this is a bug that should be fixed.

**Comment by jurgenhaas** [comment]
Thanks you @ktsaou and let me just say that I'm super impressed about your responsiveness in this issue queue.

**Comment by ktsaou** [comment]
> Thanks you @ktsaou and let me just say that I'm super impressed about your responsiveness in this issue queue.

I hate bugs, so I try to find them everywhere and solve them asap.

Actually, I would love to have to dedicated team working around the clock, on github, supporting users and extending netdata. This would also be perfect when we will have public health monitoring servers (like the global registry). It would allow us to integrate chat bots to it, or other interactive means to get in touch with your netdata servers. I strongly believe this is the way monitoring should be done these days.

Of course, as with all open source projects, the only problem is funding it...

**Comment by jurgenhaas** [comment]
This is so true. And funding is somehow getting people to contribute. But everyone can only provide certain portion of their working hours to open source projects and there are so many of them. In fact, all my business is already based on open source solutions and I give back around 20% of my time.

However, NetData is certainly on the right track. I've had many discussions with the guy who started ServerDensity a few years back and I brought all my clients as one of their first users. But he is not into that open source idea at all which is why I switched over to NetData. All the activities here are very encouraging and I hope that you will manage to build a community around it such that more people are pushing in the same direction.

**Comment by ktsaou** [comment]
I think that for the time being, the key contribution required is: **speak up**:

1. If you love it, **spread the word** (give the project a star, post on facebook, twitter, or whatever)
2. If you found a bug or something that is not working well, **open a github issue**
3. If you hate it, **let us know why**

I just monitor the badges on the wiki home. If we reach 10x the daily growth rate (impossible I think), or 40x the installed base (quite possible), I believe there will be funding available...

**Comment by ktsaou** [comment]
I updated the [Can I help?](https://github.com/firehol/netdata/wiki#can-i-help) section on the wiki to reflect the above...

**Comment by paulfantom** [comment]
Every object (module instance) is storing its timetable (last run time, next run time, and interval) in a variable `self.timetable`. In `base.py` there is a method `_run_once` which is responsible for timing (calculating intervals) and modifying `self.timetable`. However it will modify  `self.timetable['last']` only if `self.update()` returns `True` (which will be returned only if module gathered some data). 
Basically the only way to have wrong time calculations is to return `True` in module `update` method and don't send any data to main netdata application. But I don't see how this can be possible.

@jurgenhaas is there any way you could replicate this and post output of `python.d.plugin debug redis`?

**Comment by jurgenhaas** [comment]
It's not possible to reproduce this on purpose but I could debug some of the hosts permanently for Redis and memcached plugins if that helps. Do you need the debug output for exactly when the alarm get's triggered or is it sufficient to get the debug output afterwards while the alarms is still active?

**Comment by paulfantom** [comment]
> Do you need the debug output for exactly when the alarm get's triggered or is it sufficient to get the debug output afterwards while the alarms is still active?

It would be great to get both, but it should be sufficient if you give me either one.

**Comment by jurgenhaas** [comment]
As I had not seen any such false positives in the last 24 hours I was almost giving up. But then I got just one on a single host and then ran immediately this: `jurgenhaas@aweb1~> /usr/bin/python /usr/libexec/netdata/plugins.d/python.d.plugin debug memcached >memcached.debug`

I got output on the console and in the file:

- [console.txt](https://github.com/firehol/netdata/files/495272/console.txt)
- [memcached.debug.txt](https://github.com/firehol/netdata/files/495271/memcached.debug.txt)

Hope that helps?

**Comment by jurgenhaas** [comment]
Unfortunately I'm still seeing these false positives with services like Redis, Memcahed and Apache. It's happening rarely but still annoying. And also these alarms once triggered, never get cleared. Although the services are running just fine.

**Comment by ktsaou** [comment]
@paulfantom how do you suggest to trace this?

What I think is happening is this:

- `python.d.plugin` is appending a time in microseconds to the `BEGIN` line. This should be the time since the last such line sent to netdata.

- For some reason, `python.d.plugin` is not reporting the right time for these modules. I can think of 2 possible reasons:

  1. The plugin tried to connect to these services and it failed. The next second it succeeds, but the time sent to netdata is not the whole duration since the last successful time, but the time since the failed attempt.

  2. The plugin is loosing a few microseconds per update. When the cumulative time is above 5 seconds, netdata triggers the alarm.

I have not seen this case on other systems / charts, so my guess is that we experience case 1 above.

We have 2 possible fixes:

a. trace this down and fix it

b. once every hour or so, pass the time `0` to `BEGIN`. This will make netdata reset the charts to current time. I don't like this, but anyway it can be a work-around until we find enough info to fix it.

What do you think?

**Comment by ktsaou** [comment]
I just saw this on squid too. In my case, the same python.d.plugin monitors mysql, sensors, etc, but these not behind.

So, `squid`, `memcached`, `redis`. Let me check the code...

... and I believe I found it....

@paulfantom check this line: https://github.com/firehol/netdata/blob/453e1d01fd32ce5ff1948e202669f9d71a827d47/python.d/python_modules/base.py#L158

1. `self.timetable['last']` is updated unconditionally to the current date/time.
2. `self.timetable['last']` is used by `self._run_once` to calculate `since_last`, which is sent to netdata here: https://github.com/firehol/netdata/blob/453e1d01fd32ce5ff1948e202669f9d71a827d47/python.d/python_modules/base.py#L143

So, the fix is simple. `self.timetable['last']` should only be set to the current time, when the update is successful. Not always.

@paulfantom could you please fix it?

**Comment by paulfantom** [comment]
This line: https://github.com/firehol/netdata/blob/453e1d01fd32ce5ff1948e202669f9d71a827d47/python.d/python_modules/base.py#L158 is executed only once on module start and it only sets initial value. Look at `_run_once` method, it only updates `self.timetable['last']` when everything was ok. If something wrong happens it returns from this method and `self.timetable['last']` is unchanged.

I will try to trace and fix it during weekend.
@ktsaou do you want me to implement a quick and dirty fix of resetting to zero every hour?

**Comment by ktsaou** [comment]
@paulfantom may I am confused. but what I see is different. The 3 lines just below the one you posted, run `self._run_once()` forever in an infinite loop.

**Comment by ktsaou** [comment]
You are right. But there is a bug somewhere... Can we emulate a few failures to trace it?

**Comment by ktsaou** [comment]
Another possibility that could lead to the same result, is the `update()` call to take quite some time. But this should be reported at the netdata python monitoring charts, which for my squid (which still remains with the alarm on) is not the case. It reports 10-20 ms.

**Comment by paulfantom** [comment]
Currently I am leaning towards an idea that plugin is loosing some time during updates. However I cannot be sure right now. 

This weekend I will have some free time to spend on netdata (first one since last couple months) so I will try to debug this and other bugs which are assigned to me.

**Comment by ktsaou** [comment]
ok, if you believe you can do it this weekend, we can wait.
The other solution is to send a zero, only at the first `BEGIN` after an `update()` failed.

**Comment by ktsaou** [comment]
@paulfantom did you find the time to check it?

**Comment by ktsaou** [comment]
With PR #1221 netdata does not anymore trust its plugins for the charts timing. It uses their reported timings only when these timings:

1. do not send the time to the future
2. do not leave the time behind by more than 20% since the last data collection

So, if data are collected, charts are never left behind.

**Comment by jurgenhaas** [comment]
Don'T know if you receive this for closed issues but I'll give it a try: I've updated all my hosts to the latest version this morning hoping that the merged code would resolve this issue but it got worse. Memcached triggers these alerts again and it looks like memcached is just working fine.

**Comment by ktsaou** [comment]
@jurgenhaas did this happen once just after the new version of netdata was started, or is it happening while netdata runs?

**Comment by ktsaou** [comment]
It would be very helpful, if you could give me the relative alarm log. You can filter the memcached alarms and export them to CSV from the dashboard.

**Comment by jurgenhaas** [comment]
Started netdata process: 9:15am
Alert turned yellow: 10:05am
Alert turned yellow: 10:06am

[netdata_alarm_log.csv.txt](https://github.com/firehol/netdata/files/578197/netdata_alarm_log.csv.txt)

**Comment by ktsaou** [comment]
hm... I think I found it. Did it turn green again? After how much time? The log does not have the green status.

**Comment by jurgenhaas** [comment]
No, it never turned back to green.

**Comment by ktsaou** [comment]
> No, it never turned back to green.

ok. Then the memcached charts should be behind for the seconds shown at the alarm. So, can you verify that the timestamp shown at a memcached chart legend (that is autorefreshing) is really left behind the other ones?

**Comment by jurgenhaas** [comment]
You're right, see screenshot:

![2016-11-08 16 34 46 screen](https://cloud.githubusercontent.com/assets/1475675/20105284/53fe1f20-a5d1-11e6-96a9-ad590d6f83fd.jpg)

**Comment by ktsaou** [comment]
So, the alarm is valid. It really stopped collecting data from memcached.

So, we have a different bug. Alarms work as expected, but somehow the memcached python plugin of netdata stopped collecting data.

Right?

If yes, can you check your error.log for memcached errors?

**Comment by jurgenhaas** [comment]
here is what I can find in the error.log:

```
error.log:2016-11-08 09:17:06: python.d ERROR: memcached_localipv6_localipv6 no data received
error.log:2016-11-08 09:17:06: python.d ERROR: memcached_localipv6_localipv6 check function failed.
error.log:2016-11-08 09:17:06: python.d INFO: Disabled memcached/localipv6
error.log:2016-11-08 09:17:06: python.d INFO: Disabled memcached/localhost
error.log:2016-11-08 09:17:14: python.d ERROR: memcached_local [Errno 32] Broken pipe
error.log:2016-11-08 09:17:14: python.d INFO: Disabled memcached/local
error.log:2016-11-08 09:18:04: python.d ERROR: memcached_localipv6_localipv6 no data received
error.log:2016-11-08 09:18:04: python.d ERROR: memcached_localipv6_localipv6 check function failed.
error.log:2016-11-08 09:18:04: python.d INFO: Disabled memcached/localipv6
error.log:2016-11-08 09:18:04: python.d INFO: Disabled memcached/localhost
error.log:2016-11-08 09:18:24: netdata: ERROR: Health alarm 'memcached_local.cache.out_of_cache_space_time': failed to evaluate calculation with error: [ $available = 64 ] [ $cache_fill_rate = 0 ] ; failed to evaluate expression with error 104 (computed value is infinite)
error.log:2016-11-08 10:05:54: alarm-notify.sh: INFO: sent slack notification for: ts1 memcached_local.cache.memcached_last_collected_secs is WARNING to 'tineon-devop'
error.log:2016-11-08 10:06:54: alarm-notify.sh: INFO: sent slack notification for: ts1 memcached_local.cache.memcached_last_collected_secs is CRITICAL to 'tineon-devop'
```

As I mentioned in the original post, memcached on this host is running but not utilized by any application, that's why its cache is empty. Might that be the reason for the alert?

**Comment by ktsaou** [comment]
hm... so it is just frozen. The python module is either connected to memcached but it does not receive a response (i.e. memcached does not respond - highly unlikely), or somehow it internally decided (due to a bug) to stop polling memcached (likely).

What I suggest is to kill `python.plugin`. netdata will restart it.

@paulfantom we really need your help here. All these are too deep for my python knowledge, so I would appreciate your help...

**Comment by ktsaou** [comment]
@jurgenhaas since it seems @paulfantom is busy, I tried to give a solution myself.

I have PR https://github.com/firehol/netdata/pull/1230 in which I set a timeout of 5 seconds to receive data from memcached. I also added some more debugging info and error messages to help us trace the issue.

If it possible to install `https://github.com/ktsaou/netdata` and try it?

**Comment by jurgenhaas** [comment]
@ktsaou I have installed your fork on 16 hosts of one of my clients just now and watch this for the next couple of hours. Will let you know what's going to happen. Thanks for your effort.

**Comment by jurgenhaas** [comment]
@ktsaou This is looking really good so far. Memcache is working just fine, no alerts and the chart is never behind.

Looks like your modification would be great for all sorts of plugins, because I got such alerts today for Redis and Apache. And I see from the code that those have not been modified. Of course, it was just a test on memcached, I guess.

**Comment by ktsaou** [comment]
I think this change affects memcached, redis and squid. I'll check apache, but I have never seen a false alarm on apache. Are you sure about it?

Also, I would love to find out why it is working now. Since I have added more error logs for that case, could you please post again the memcached logs from `error.log`?

**Comment by ktsaou** [comment]
> Also, I would love to find out why it is working now. Since I have added more error logs for that case, could you please post again the memcached logs from error.log?

my bad. you will not find any logs. I added `debug` statements that are logged only when the plugin runs with debug enabled.

**Comment by ktsaou** [comment]
So, tell me more about your `redis` and `apache` false alarms.
`redis` should be solved with this version. Do you have any false alarms with it?

**Comment by ktsaou** [comment]
I merged the last fix.

**Comment by jurgenhaas** [comment]
Sorry for my late reply on this. Here are more details about redis and apache:

Redis had been my mistake yesterday, what I saw was an old one from the night before. However, I got one this morning which hasn't turned green yet although Redis is working just fine on that host. The alarm export CSV is attached. And a portion of the error.log around that time is also attached, which doesn't seem to contain much more data.

The apache error was in the afternoon yesterday but I had to reboot that host this morning and can't reproduce what was the problem yesterday, so let's ignore that for now.

[netdata_alarm_log_redis.csv.txt](https://github.com/firehol/netdata/files/583093/netdata_alarm_log_redis.csv.txt)
[netdata.error.log.txt](https://github.com/firehol/netdata/files/583099/netdata.error.log.txt)

**Comment by ktsaou** [comment]
Could you please verify that the redis chart `redis_local.operations` is not 60+ seconds behind other non-redis charts or the current time? This is very important:

- If it is 60+ seconds behind the problem is at the data collection timing.
- If is not 60+ seconds behind and the alarm is raised, the problem is at the alarms

**Comment by jurgenhaas** [comment]
Of course, you're right. Just checked and the charts are 3 hours behind.

**Comment by ktsaou** [comment]
So it is a data collection problem. is the new version running on them? If yes, my fix was not good enough...

**Comment by ktsaou** [comment]
Based on your report I understand that the charts that are left 3 hours behind, are collecting new data. Right?

If yes, are you sure these netdata are recent? I remind you that recently we have applied 2 fixes:

1. charts that are collecting new data, now should never stay behind in time, no matter what the plugins do. So far, I believe this works perfectly. If yours is using the latest netdata, collects new data and shows a time 3 hours back, it will be the first.

2. redis, memcached and squid (SocketService based python.d.plugin modules) should not freeze/block. So far, I believe this also works perfectly. If yours is using the latest netdata and redis is frozen, it will be the first and I'll need to investigate further.

**Comment by jurgenhaas** [comment]
Sorry i wasn't clear about this. The charts are behind and do not collect data. Their charts are standing still.

**Comment by ktsaou** [comment]
And latest version of netdata? We didn't fix it?

**Comment by ktsaou** [comment]
I did another re-work on the ktsaou fork. I would appreciate if you could test it on a redis. This time, I reworked quite a bit of the SocketService and I added error logging in case either abnormal happens to the SocketService connection.

**Comment by jurgenhaas** [comment]
Thanks @ktsaou I've just installed that fork and will watch it during the day.

**Comment by jurgenhaas** [comment]
@ktsaou Everything is running so smoothly today, it's almost unbelievable. Looks like you've nailed it.

**Comment by ktsaou** [comment]
nice! I have done a few changes and merged them. I also added a few more charts to redis and changed the way memcached is presented (names, titles, colors, etc).
Give them a try... (already merged into main repo).

**Comment by jurgenhaas** [comment]
I hate to say this but I'm still getting these alerts on different hosts. This happened for redis, mysql and apache this morning and in all cases

- the charts on the dashboard are standing still, i.e. they are behind the current time and not collecting data anymore
- the services are running ok
- the alerts are not returning to green
- the error.log doesn't show anything

An interesting observation might be the following screenshot:

![2016-11-14 11 34 42 screen](https://cloud.githubusercontent.com/assets/1475675/20261533/76a0d342-aa5e-11e6-89fe-af0ff0b6cce9.jpg)

It shows for apache.local that it stopped collecting data at around 11:30 and right above that one, the user.net chart shows that at around the same time, there has been an interruption, i.e. no data for a few seconds, but that recovered from that scenario where as apache.local didn't.

**Comment by ktsaou** [comment]
hm... this is totally strange.

I think there should be errors logged in netdata.conf for these now. Could you please check?

To be sure everything is properly controlled, I ended up rewriting most of the timing part in python, with PR #1248

So please try this one.
(I am very sorry for this try-and-fail game - I really hope @paulfantom is ok - so you have just me for the moment).

**Comment by jurgenhaas** [comment]
@ktsaou no worries, I'm perfectly OK going through this phase as I can certainly see that this helps this amazing product.

And the good news so far, no false alarms for the last 4 hours. Let's wait with the celebreation a little longer, but I'm hopeful.

**Comment by jurgenhaas** [comment]
It has been so silent today that I was wondering if the alarms were still working ;-)

I wanted to force an alert on one host and killed redis-server on it - which immediately triggered the alert. Then I also killed apache and got the second alert. When I started apache again, the number of alerts got reduced to one which is great. Then I started redis-server agin, but that was then no longer picked up by netdata. The alert had turned to red meanwhile and it did not recoved from that.

Other than that, I continue to watch all this on 30+ hosts and will let you know tomorrow how it goes.

**Comment by ktsaou** [comment]
> Then I started redis-server agin, but that was then no longer picked up by netdata. The alert had turned to red meanwhile and it did not recoved from that.

python.d.plugin retires after 60 seconds, by default. So, if you stop a service, python.d.plugin will check it 60 times and then forget about it. This is controlled by its config. I think though the default should probably be 600. What do you think?

**Comment by LeeNX** [comment]
Just my 2c, how can you get a server/service back into the monitor rotation without restarting netdata?  
  
Also, is there some retry notation that could be used, that would have netdata retries get incrementally/logarithmically get longer, ie 60x1s, 120x5s, 240x30s, etc ... This might be a neat way to have repaired services reappear?

**Comment by ktsaou** [comment]
ok, this is what I did:

In case of failures, python.d.plugin calculates a `penalty` which it add to each iteration, until it managed to successfully collect data again:

1. for the first `retries` (by default `60`), `penalty = 0`. So, it does what it already does today.
2. for the next `retries`, `penalty = retries * update every / 2`, so for per-second data collection and 60 retries, it will do 60 more tries with 30 seconds between each retry.
3. for the next `retries` it will use `penalty = penalty * 1.5`, so 60 more tries with 45 seconds between all tries.
4. it will continue with 3, increasing `penalty` by `1.5` every time, until `penalty` reaches 600 seconds (10 minutes). At this point it will continue doing one check every 10 minutes forever.

How does that sound?

**Comment by LeeNX** [comment]
@ktsaou - Don't you ever sleep? ;-)  
  
Above sounds very sane.

**Comment by ktsaou** [comment]
haha... merged that one too.

**Comment by jurgenhaas** [comment]
This is looking very good and is in action here already. Also, my test scenario from yesterday is now working nicely. Great job @ktsaou !

**Comment by ktsaou** [comment]
ok. I am closing this.
Keep in mind I found another case where python.d.plugin was retiring, which I fixed.
So, it might be good to use the latest version.
